### DISCUSSION TITLE
Scheduler change with 3.6.5? Schedule for every minute is not working anymore

### DISCUSSION THREAD

**Discussion by eh-gh** [emsesp/EMS-ESP32#1740] [Q&A]
Hi here,

I'm using a scheduler entry to check every minute a water level sensor. If the level has reached I switch on a water pump connected to another pin.
It does not work after upgrade from 3.6.4 to 3.6.5 (and with the latest dev release either).

Evidence is attached.
Can you check plz?

tnx
eh

[emsesp_customizations.json](https://github.com/emsesp/EMS-ESP32/files/15210051/emsesp_customizations.json)
[emsesp_schedule.json](https://github.com/emsesp/EMS-ESP32/files/15210052/emsesp_schedule.json)
[latest_dev_log.txt](https://github.com/emsesp/EMS-ESP32/files/15210053/latest_dev_log.txt)
[working_3.6.4_log.txt](https://github.com/emsesp/EMS-ESP32/files/15210060/working_3.6.4_log.txt)

ps.
the mqtt error in dev log can not be seen in log 3.6.5, but the scheduler does not work with that version either.

**Comment by MichaelDvP** [comment]
Seems you are running out of memory. Please attach the support information.

**Comment by eh-gh** [comment]
here you are
fyi: the version 3.6.5 does not show the "out of memory" but the scheduling problem exists.

tnx

[emsesp_system_info.txt](https://github.com/emsesp/EMS-ESP32/files/15210445/emsesp_system_info.txt)

**Reply by MichaelDvP** [reply]
Looks ok. Are you sure the command is not executed? We have changed the log level for the success message to debug.
https://github.com/emsesp/EMS-ESP32/blob/43777f4a151702d1231a33ab2f8b1a747f800716/src/web/WebSchedulerService.cpp#L349
so you do not see it.

**Reply by proddy** [reply]
@eh-gh set your log level in the WebUI to ALL in System->Log and see if the Scheduler command is actually being called.

**Comment by eh-gh** [comment]
just tested with the some water: as you can see the water level has reached but the pump does not start
in case I would fall back to 3.6.4 the scheduler works.

![Screenshot_20240504_204049_com android chrome](https://github.com/emsesp/EMS-ESP32/assets/77990285/c303d69e-5350-430c-92b0-b26547033127)

**Reply by MichaelDvP** [reply]
Found the bug. It's not the scheduler, it's calling a command with command as value. Fixed in my dev.

**Reply by proddy** [reply]
great Michael! I saw your fix. It must have been broken for months

**Reply by MichaelDvP** [reply]
Since ArduinoJson 7, it's also broken in 3.6.5 main.

**Reply by proddy** [reply]
shall I make a 3.6.6 patch?

**Reply by MichaelDvP** [reply]
BTW: the out_of_memory mqtt message is from sending the HA config. The boiler now has dhw nested and HA-config tries to send all entities at once. It stops on the memory limit and continues in the next mqtt send, there is no message lost.
To avoid the message mqtt can set to individual topics, that split the boiler message in biler and boiler_dhw  again.

**Reply by proddy** [reply]
I prefer to keep the nested and not split topics if possible. There must be another way, is it the `publish_ha_sensor_config()` ?

**Reply by eh-gh** [reply]
just verified the original problem with 3.7.0-dev.10
it works again
thanks for the amazing fast correction

**Reply by MichaelDvP** [reply]
> shall I make a 3.6.6 patch?

I don't think this feature is used often, a patch is nice if it's not too much affort.
>  There must be another way,

We can limit the number of HA configs send at once. After start the telegrams come in one by one and HA config is created on each mqtt publish a small number of new entities. My log with counter and 10 sec nested mqtt publish shows:
```
2024-05-05 11:09:15.966 I 15: [emsesp] Recognized new boiler with deviceID 0x08
2024-05-05 11:09:16.033 I 16: [emsesp] Created 5 HA-boiler-configs
2024-05-05 11:09:19.911 I 17: [emsesp] Recognized new controller with deviceID 0x09
2024-05-05 11:09:20.208 I 18: [emsesp] Recognized new connect with deviceID 0x0B
2024-05-05 11:09:20.726 I 19: [emsesp] Recognized new thermostat with deviceID 0x10
2024-05-05 11:09:21.154 I 20: [emsesp] Recognized new mixer with deviceID 0x21
2024-05-05 11:09:21.571 I 21: [emsesp] Recognized new solar with deviceID 0x30
2024-05-05 11:09:26.002 I 22: [emsesp] Created 62 HA-boiler-configs
2024-05-05 11:09:26.325 I 23: [emsesp] Created 14 HA-solar-configs
2024-05-05 11:09:36.058 I 26: [emsesp] Created 45 HA-thermostat-configs
2024-05-05 11:09:46.059 I 27: [emsesp] Created 4 HA-thermostat-configs
2024-05-05 11:09:46.237 I 28: [emsesp] Created 8 HA-solar-configs
2024-05-05 11:09:46.494 I 29: [emsesp] Created 2 HA-mixer-configs
2024-05-05 11:09:56.003 I 31: [emsesp] Created 10 HA-boiler-configs
2024-05-05 11:09:56.239 I 32: [emsesp] Created 3 HA-thermostat-configs
2024-05-05 11:09:56.519 I 33: [emsesp] Created 4 HA-mixer-configs
2024-05-05 11:10:06.062 I 34: [emsesp] Created 7 HA-thermostat-configs
2024-05-05 11:10:16.001 I 35: [emsesp] Created 3 HA-boiler-configs
```

eh-gh have 77 boiler entities including dhw, heatpumps have over 100 without dhw, With esp without psram they will see this memory message also with older versions and boiler_ww mqtt.
As mentioned it is no issue, on the memory limit mqtt stops generating HA messages and continue on next publish, no config message is lost.

**Reply by MichaelDvP** [reply]
Oh, we have 2 different memory checks, we should sync them. 
https://github.com/emsesp/EMS-ESP32/blob/8084b9141d1f24ae8c86234dfa9b7a31f9271bd9/src/emsdevice.cpp#L1771
is the check for publish, if fail the errormessage is created here:
https://github.com/emsesp/EMS-ESP32/blob/8084b9141d1f24ae8c86234dfa9b7a31f9271bd9/src/mqtt.cpp#L611-L617
but the HA ggeneration also should break a bit before, but with different test:
https://github.com/emsesp/EMS-ESP32/blob/8084b9141d1f24ae8c86234dfa9b7a31f9271bd9/src/emsdevice.cpp#L1777
I think we should always check the heap_caps with 60k limit in mqtt and 65k or 70k in emsdevice. So we can always add a large boiler_data message without getting to the mqtt limit and create the remaining HA configs later.

**Reply by proddy** [reply]
thanks @MichaelDvP for fixing the HA memory issue
